### PR TITLE
PERF: Cache each theme field value once

### PR DESCRIPTION
**Pull request by danielwaterworth**
Previously, theme fields from components would be cached for each of their parent themes.

### CODE REVIEW

**Comment by tgxworld** [comment]
I think more context needs to be provided in order to make this PR easier to review for a reviewer. Here are some questions that I had when I looked at this PR. 

1. What is the problem are we trying to solve here? 
1. What is the solution that is being implemented here? 
1. What is the performance benefit we get as a result of this solution being implemented? Is it smaller cache size?

File: app/models/theme.rb [hunk 1]
```diff
@@ -434,25 +429,71 @@ def self.refresh_message_for_targets(targets, theme_ids)
...
+  end
+
+  def self.theme_field_values(theme_ids, targets, names)
+    cache.defer_get_set_bulk(
+      Array(theme_ids).product(Array(targets), Array(names)),
+      lambda do |(theme_id, target, name)|
+        "#{theme_id}:#{target}:#{name}:#{Theme.compiler_version}"
+      end,
+    ) do |keys|
+      keys.map! { |theme_id, target, name| [theme_id, Theme.targets[target], name.to_s] }
```
Comment (tgxworld) [inline app/models/theme.rb:484]: I'm wondering if this is intentional where the keys being generated based on the inputs to `defer_get_set_bulk` can be mutated by the block.

**Comment by tgxworld** [comment]
I think it'll be useful to add some test cases here even though there are none at the moment. While reviewing the PR, I wrote one which I think can be incorporated to this PR with the right assertions:

```
  context "when theme component is used in multiple themes" do
    fab!(:theme_1) { Fabricate(:theme, user: user) }
    fab!(:theme_2) { Fabricate(:theme, user: user) }

    it "efficiently caches fields of theme component by only caching the fields once across multiple themes" do
      child.set_field(target: :common, name: "header", value: "World")
      child.save!

      theme.add_relative_theme!(:child, child)
      theme_2.add_relative_theme!(:child, child)

      expect(Theme.lookup_field(theme.id, :desktop, "header")).to eq("World")
      expect(Theme.lookup_field(theme_2.id, :desktop, "header")).to eq("World")

      puts Theme.cache.hash
    end
  end
```

For the test case above, the things we stored in the cache changed from

```
"905,904:desktop:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>"World"
"903,904:desktop:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>"World"
```
to 

```
"901:0:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>nil, "901:1:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>nil, "900:0:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>["World"], "900:1:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>nil,
"899:0:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>nil, "899:1:header:2c9a63bcd0caba1ae77eb389b6eb30f80021836f"=>nil
```

and we increases the number of keys used in cache and I wonder if it'll affect the work you're doing to limit the number of items in our in process cache.

File: app/models/theme.rb (hunk 1 above)
Comment (danielwaterworth) [reply app/models/theme.rb:484]: Oh, good catch.

**Review by tgxworld — APPROVED** [review]

**Comment by tgxworld** [comment]
Thank you for the changes :+1: It looks good to me.
